### PR TITLE
LibJS: Convert resolve_this_binding and to_reference to ThrowCompletionOr

### PR DESCRIPTION
**Pull request by davidot**
The `to_reference()` change touches AST a lot so maybe this should be combined to not cause too many conflicts?

### CODE REVIEW

**Comment by linusg** [comment]
Don't worry about conflicts, I'm happy to fix them :^)

File: Userland/Libraries/LibJS/AST.cpp [hunk 1]
```diff
@@ -682,11 +680,11 @@ struct ForInOfHeadState {
                 if (is<VariableDeclaration>(*expression_lhs)) {
                     auto& declaration = static_cast<VariableDeclaration const&>(*expression_lhs);
                     VERIFY(declaration.declarations().first().target().has<NonnullRefPtr<Identifier>>());
-                    lhs_reference = declaration.declarations().first().target().get<NonnullRefPtr<Identifier>>()->to_reference(interpreter, global_object);
+                    lhs_reference = TRY(declaration.declarations().first().target().get<NonnullRefPtr<Identifier>>()->to_reference(interpreter, global_object));
                 } else {
                     VERIFY(is<Identifier>(*expression_lhs) || is<MemberExpression>(*expression_lhs));
                     auto& expression = static_cast<Expression const&>(*expression_lhs);
-                    lhs_reference = expression.to_reference(interpreter, global_object);
+                    lhs_reference = TRY(expression.to_reference(interpreter, global_object));
```
Comment (linusg) [inline Userland/Libraries/LibJS/AST.cpp:687]: With this the exception check on line 710 should no longer be needed as everything else can't throw or uses `MUST`

File: Userland/Libraries/LibJS/AST.cpp [hunk 2]
```diff
@@ -1237,17 +1233,15 @@ Value UnaryExpression::execute(Interpreter& interpreter, GlobalObject& global_ob
 
     auto& vm = interpreter.vm();
     if (m_op == UnaryOp::Delete) {
-        auto reference = m_lhs->to_reference(interpreter, global_object);
+        auto reference = TRY_OR_DISCARD(m_lhs->to_reference(interpreter, global_object));
         if (interpreter.exception())
```
Comment (linusg) [inline Userland/Libraries/LibJS/AST.cpp:1237]: exception check not needed anymore

File: Userland/Libraries/LibJS/AST.cpp [hunk 3]
```diff
@@ -1528,7 +1522,7 @@ ThrowCompletionOr<Value> ClassExpression::class_definition_evaluation(Interprete
 
         Value super_class;
 
-        auto reference = m_super_class->to_reference(interpreter, global_object);
+        auto reference = TRY(m_super_class->to_reference(interpreter, global_object));
         if (auto* exception = interpreter.exception())
```
Comment (linusg) [inline Userland/Libraries/LibJS/AST.cpp:1526]: exception check not needed anymore

**Review by linusg — COMMENTED** [review]
Looks great, thanks! I *think* we can get rid of a couple more manual exception checks.

**Comment by davidot** [comment]
Remove those no longer needed exception checks.
Ran test262 and it shows no changes so seems good.

**Review by linusg — APPROVED** [review]

**Pull request merged**